### PR TITLE
[CHM 289 ] km-add-file component

### PR DESCRIPTION
**Pull request by liyuanmontreal**
Component composition
Similar to km-add-link, km-add-file contains
a select-file-button,
a file-upload-editor,
and also use km-view-links component.

Future work
remain work:  uploading file and showing uploading progress or error info on file-upload-editor.

Other change
This PR also contains minor changes in km-view-links for improve list style.

### CODE REVIEW

File: src/components/controls/link/km-add-file.vue [hunk 1]
```diff
@@ -0,0 +1,69 @@
+<template>    
+    <div class="scbd-common km-add-file" >
+        <select-file-button  :multiple="props.multiple" @on-file-selected="receiveFile"> <slot name="file-button-label">+ Add Files</slot></select-file-button> 
+              
+        <file-upload-editor ref="fileEditorRef" @on-close="onFileUploadEditorClose">
+            <template v-slot:modalTitle>
+            <slot name="link-dialog-title"   >        
+                Uploading file
```
Comment (blaisefonseca) [inline src/components/controls/link/km-add-file.vue]:
```suggestion
     File upload   
```

File: src/components/controls/link/km-add-file.vue [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+            </slot>
+            </template>       
+        </file-upload-editor>
+        <slot name="links-view">
+            <km-view-links v-model="links"   @on-delete = "removeLink"   @on-edit = "editLink"></km-view-links>  
+        </slot>
+    </div>
+</template>
+<script setup>
+    import _ from "lodash";
```
Comment (blaisefonseca) [inline src/components/controls/link/km-add-file.vue]: Import only functions which are used for eg
Import { map} from lodash

File: src/components/controls/link/km-add-file.vue [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+            else {        
+                links.value[editedLinkIndex]=newValue;
+            } 
+            editedLinkIndex = -1;
+        }       
+    }
+
+    const files = ref([]);
+    const receiveFile = (receiveFiles) => { 
+        receiveFiles = _.flatten([receiveFiles]); // systematic array
```
Comment (blaisefonseca) [inline src/components/controls/link/km-add-file.vue]: You can use native function on array receiveFiles.flat()

File: src/components/controls/link/km-view-links.vue [hunk 4]
```diff
@@ -2,12 +2,24 @@
...
+                        {{item.name}} 
+                    </a>
+                </td>  
+                <td class="align-middle">                   
+                    <span v-if="item.tag">
+                        ({{ item.tag }})
+                    </span>            
+                </td>  
+                <td class = "text-end align-middle">
+                    <span class="btn" @click="edit(index)"><i class="bi bi-pencil-square"></i></span>
```
Comment (blaisefonseca) [inline src/components/controls/link/km-view-links.vue]: Can you verify if the abs project is using the same format for display?

**Review by blaisefonseca — COMMENTED** [review]
Please check comments

File: src/components/controls/link/km-add-file.vue (hunk 3 above)
Comment (liyuanmontreal) [reply src/components/controls/link/km-add-file.vue]: fixed

File: src/components/controls/link/km-add-file.vue (hunk 2 above)
Comment (liyuanmontreal) [reply src/components/controls/link/km-add-file.vue]: fixed

File: src/components/controls/link/km-add-file.vue [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+        }       
+    }
+
+ 
+    const receiveFile = (receiveFiles) => {  
+        receiveFiles = [receiveFiles];
+        receiveFiles = receiveFiles.flat();   
+        //TODO: support multiple files   
+        addLink(receiveFiles[0]);            
+    };
```
Comment (blaisefonseca) [inline src/components/controls/link/km-add-file.vue]:
```suggestion
    const receiveFile = (selectedFile) => {  
         addLink(selectedFile);            
    };
```

File: src/components/controls/link/km-add-file.vue [hunk 6]
```diff
@@ -0,0 +1,67 @@
+<template>    
+    <div class="scbd-common km-add-file" >
+        <select-file-button  :multiple="props.multiple" @on-file-selected="receiveFile"> <slot name="file-button-label">+ Add Files</slot></select-file-button>
```
Comment (blaisefonseca) [inline src/components/controls/link/km-add-file.vue]:
```suggestion
        <select-file-button  @on-file-selected="receiveFile"> <slot name="file-button-label">+ Add Files</slot></select-file-button> 
```

**Review by blaisefonseca — APPROVED** [review]

**Pull request merged**